src/backend/distributed/commands/table.c

```diff
@@ -214,13 +214,19 @@ PostprocessCreateTableStmtPartitionOf(CreateStmt *createStatement, const
 
 	Assert(parentRelationId != InvalidOid);
 
+	Oid relationId = RangeVarGetRelid(createStatement->relation, NoLock, missingOk);
+
 	/*
 	 * If a partition is being created and if its parent is a distributed
 	 * table, we will distribute this table as well.
+	 * In case of an IF NOT EXISTS statement, we also should make sure that:
+	 * the relation is a partition, the partition is not already distributed,
+	 * and that partition's parent is the current parent from parentRelationId
 	 */
-	if (IsCitusTable(parentRelationId))
+	if (IsCitusTable(parentRelationId) && !IsCitusTable(relationId) &&
```

Comment:
I think in terms of correctness, this is a pretty comprehensive check, preventing several edge cases, which is great. (Also, great test coverage!)

In terms of the code, I see some potential improvement areas, please feel free to consider the following.

I think with this check, we are trying to do multiple unrelated checks at once and also behaving `if_not_exists`  very implicitly. 

Instead, it'd be easier to follow if we make these checks explicitly. Please see the diff: https://github.com/citusdata/citus/compare/issue2087...onder_issue2087?expand=1

I think that you don't have to (and probably shouldn't as the comments have to be be improved) the proposed diff above, but something along with these lines can be nicer in terms of code readability.

Reply:
It seems more readable to completely separate the IF NOT EXISTS case. Will go for it :+1: